pubsub/gossipsub/gossipsub-v1.3.md

```diff
@@ -0,0 +1,282 @@
+# gossipsub v1.3: Choke extensions to improve network efficiency
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2025-05-23  |
+
+Authors: [@marcopolo]
+
+Interest Group: TODO
+
+[@marcopolo]: https://github.com/marcopolo
+
+See the [lifecycle document][lifecycle-spec] for context about the maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+## Overview
+
+This version specifes two protocol changes to gossipsub v1.2:
+
+- Two new control messages: Choke/Unchoke.
+- Mesh peers MAY send IHAVEs in lieu of pushing a message. This is typically
+  done in response to being Choked.
+
+The two new control messages, `Choke` and `Unchoke`, control whether mesh peers
+will eagerly or lazily push messages. A peer is considered choked from the
+perspective of another peer if it has received a `Choke` message from that peer.
+A choked peer is unchoked if it has received an `Unchoke` message from that
+peer, or if it leaves that peer's mesh. Peers are initially unchoked when
+grafted to a mesh. Choked status is not automatically symmetric. For peers A and
+B, A may choke B, while B may have A unchoked.
+
+If there are no choked peers in the mesh, this version of gossipsub behaves
+identically to the previous version of gossipsub. A mesh with only choked peers
+behaves identically to a mesh with no chocked peers with an additional
+network round trip of latency when fetching the message payload. Latency may
+increase further if the peer limits the number of concurrent IWANTs.
+
+When choking is used well, messages arrive without extra delay and without
+excessive duplicates. The graph of unchoked peers naturally evolves to utilize
+better network paths.
+
+## Terms and Definitions
+
+Eager Push: A message is sent to a peer immediately, without receiving a prior
+`IWANT` request.
+
+Lazy Push: A message ID is sent to a peer in an `IHAVE` rather than sending the
+message itself. A peer will only receive the message after explicitly requesting
+it.
+
+Choked peer: A peer that is currently choked and should lazy push rather than
+eager push.
+
+Unchoked peer: A peer that is unchoked and will eagerly push messages.
+
+Choking peer: From the perspective of a node, this is the peer that has choked
+it.
+
+## State Diagram
+
+```
+      ┌─────────────┐
+      │   GRAFTED   │
+      │  (Initial)  │
+      └──────┬──────┘
+             │
+             ▼
+      ┌─────────────┐
+┌─────│  UNCHOKED   │◀────┐
+│     │  (Default)  │     │
+│     └──────┬──────┘     │
+│            │            │
+│            │ Choke      │ Unchoke
+│            │ message    │ message
+│            ▼            │
+│     ┌─────────────┐     │
+│     │   CHOKED    │─────┘
+│     │             │
+│     └─────────────┘
```

Comment:
Should there be an arrow from CHOKED to PRUNED as well? because the node can leave the mesh when choked?